File: Programmers/67259Programmers.py
import sys
import copy
from collections import deque
sys.setrecursionlimit(2000)

dir = [[0, 1], [0, -1], [1, 0], [-1, 0]]

def bfs(board, boardCopy):
    
    maxR = len(board)
    maxC = len(board[0])
    q = deque()
    q.append({'r': 0, 'c': 0, 'cost': 0, 'dirNum': 4})
    
    while len(q):
        curr = q.popleft()
        # print(curr)
        # if nr < 0 or nc < 0 or nr >= maxR or nc >= maxC:
        if boardCopy[curr['r']][curr['c']] < curr['cost']:
            continue
        
        boardCopy[curr['r']][curr['c']] = curr['cost']
        for i in range(0, 4):
            nr = curr['r'] + dir[i][0]
            nc = curr['c'] + dir[i][1]
            if nr < 0 or nc < 0 or nr >= maxR or nc >= maxC:
                continue
            
            if board[nr][nc] == 1:
                continue
            
            if curr['dirNum'] == 4:
                if boardCopy[nr][nc] < curr['cost'] + 100:
                    continue
                q.append({'r': nr, 'c': nc, 'cost': curr['cost'] + 100, 'dirNum': i})
            elif i == curr['dirNum']:
                if boardCopy[nr][nc] < curr['cost'] + 100:
                    continue
                q.append({'r': nr, 'c': nc, 'cost': curr['cost'] + 100, 'dirNum': i})
            else:
                if boardCopy[nr][nc] < curr['cost'] + 600:
                    continue
                q.append({'r': nr, 'c': nc, 'cost': curr['cost'] + 600, 'dirNum': i})
                
    
    
def solution(board):
    boardCopy = [[987654321]*len(board[0]) for _ in range(len(board))]

    bfs(board, boardCopy)
    answer = boardCopy[len(board)-1][len(board[0])-1]
    return answer

File: Programmers/67259Programmers.cpp
// 경주로 건설
#include <string>
#include <vector>
#include <queue>
#include <iostream>

#define MAX_N 25

using namespace std;

struct pos{
    int r, c;
    int cost;
    int dirNum;
};

int G[MAX_N+1][MAX_N+1];
int dir[4][2] = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};

void bfs(vector<vector<int>> &board){
    queue<pos> q;
    q.push({0, 0, 0, 4});
    int m = board.size();
    int n = board[0].size();
    
    while(!q.empty()){
        pos curr = q.front();
        q.pop();

        if(G[curr.r][curr.c] < curr.cost){
            continue;
        }
        
        G[curr.r][curr.c] = curr.cost;
        for(int i=0; i<4; ++i){
            int nr = curr.r + dir[i][0];
            int nc = curr.c + dir[i][1];
            if(nr < 0 || nc < 0 || nr >= m || nc >= n){
                continue;
            }
            if(board[nr][nc] == 1){
                continue;
            }
            if(curr.dirNum == 4){
                q.push({nr, nc, curr.cost + 100, i});
                continue;
            }
            if(curr.dirNum != i){
                q.push({nr, nc, curr.cost + 600, i});
            } else{
                q.push({nr, nc, curr.cost + 100, i});
            }
        }
    }
}

int solution(vector<vector<int>> board) {
    int answer = 0;
    for(int i=0; i<board.size(); ++i){
        for(int j=0; j<board[0].size(); ++j){
            G[i][j] = 987654321;
        };
    }
    bfs(board);
    
    answer = G[board.size()-1][board[0].size()-1];
    
    return answer;
}